File: app/src/main/java/com/mrgames13/jimdo/feinstaubapp/Services/WebRealtimeSyncService.java
/*
 * Copyright © 2019 Marc Auberer. All rights reserved.
 */

package com.mrgames13.jimdo.feinstaubapp.Services;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.IBinder;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.mrgames13.jimdo.feinstaubapp.App.MainActivity;
import com.mrgames13.jimdo.feinstaubapp.CommonObjects.Sensor;
import com.mrgames13.jimdo.feinstaubapp.R;
import com.mrgames13.jimdo.feinstaubapp.Utils.StorageUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class WebRealtimeSyncService extends Service {

    // Variables as objects
    private static DatabaseReference ref;
    ArrayList<Sensor> favourites;
    ArrayList<Sensor> own_sensors;

    // Utils packages
    private StorageUtils su;

    // Variables
    public static WebRealtimeSyncService own_instance;
    private long timestamp;

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        String sync_key = intent.getStringExtra("sync_key");

        // Initialize own instance
        own_instance = this;

        // Initialize StorageUtils
        su = new StorageUtils(getApplicationContext());

        // Initialize Firebase
        FirebaseDatabase db = FirebaseDatabase.getInstance();
        ref = db.getReference("sync/" + sync_key);

        refresh(getApplicationContext());

        return START_NOT_STICKY;
    }

    public void refresh(final Context context) {
        timestamp = System.currentTimeMillis();

        // Get favourites and own sensors from local db
        favourites = su.getAllFavourites();
        own_sensors = su.getAllOwnSensors();

        // Assenble data
        final HashMap<String, Object> data = new HashMap<>();
        int object_id = 0;
        for(Sensor s : favourites) {
            HashMap<String, Object> sensor_map = new HashMap<>();
            sensor_map.put("name", s.getName());
            sensor_map.put("chip_id", s.getChipID());
            sensor_map.put("color", String.format("#%06X", (0xFFFFFF & s.getColor())));
            sensor_map.put("fav", true);
            data.put(String.valueOf(object_id), sensor_map);
            object_id++;
        }
        for(Sensor s : own_sensors) {
            HashMap<String, Object> sensor_map = new HashMap<>();
            sensor_map.put("name", s.getName());
            sensor_map.put("chip_id", s.getChipID());
            sensor_map.put("color", String.format("#%06X", (0xFFFFFF & s.getColor())));
            sensor_map.put("fav", false);
            data.put(String.valueOf(object_id), sensor_map);
            object_id++;
        }

        // Build connection
        HashMap<String, Object> connection = new HashMap<>();
        connection.put("time", timestamp);
        connection.put("device", "app");
        connection.put("data", data);
        ref.setValue(connection);

        // Set DateChangeListener
        ref.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snap) {
                if(snap.exists()) {
                    if(snap.child("device").getValue() != null && !snap.child("device").getValue().equals("app")) {
                        ArrayList new_sensors = (ArrayList) snap.child("data").getValue();
                        if(new_sensors != null && new_sensors.size() > 0) {
                            favourites = su.getAllFavourites();
                            own_sensors = su.getAllOwnSensors();
                            for(Sensor s : favourites) su.removeFavourite(s.getChipID(), true);
                            for(Sensor s : own_sensors) su.removeOwnSensor(s.getChipID(), true);
                            for(int i = 0; i < new_sensors.size(); i++) {
                                Map<String, Object> sensor = (Map<String, Object>) new_sensors.get(i);
                                // Extract data
                                String chip_id = String.valueOf(sensor.get("chip_id"));
                                String name = String.valueOf(sensor.get("name"));
                                boolean favorized = Boolean.parseBoolean(sensor.get("fav").toString());
                                String color = String.valueOf(sensor.get("color"));
                                if(!su.isFavouriteExisting(chip_id) && !su.isSensorExisting(chip_id)) {
                                    if(favorized) {
                                        su.addFavourite(new Sensor(chip_id, name, Color.parseColor(color)), true);
                                    } else {
                                        su.addOwnSensor(new Sensor(chip_id, name, Color.parseColor(color)), true, true);
                                    }
                                }
                            }
                            MainActivity.own_instance.pager_adapter.refreshFavourites();
                            MainActivity.own_instance.pager_adapter.refreshMySensors();
                        }
                    }
                } else {
                    ref.removeEventListener(this);
                    // Show toast
                    LayoutInflater inflater = (LayoutInflater) context.getSystemService(LAYOUT_INFLATER_SERVICE);
                    Toast t = new Toast(context);
                    t.setGravity(Gravity.CENTER, 0, 0);
                    t.setDuration(Toast.LENGTH_LONG);
                    t.setView(inflater.inflate(R.layout.sync_ended, null));
                    t.show();
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                Toast.makeText(context, "Sync failed", Toast.LENGTH_SHORT).show();
            }
        });
    }

    public void stop() {
        if(ref != null) ref.removeValue();
        stopSelf();
    }
}

File: app/src/main/java/com/mrgames13/jimdo/feinstaubapp/service/WebRealtimeSyncService.kt
/*
 * Copyright © 2019 Marc Auberer. All rights reserved.
 */

package com.mrgames13.jimdo.feinstaubapp.service

import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.IBinder
import android.view.Gravity
import android.view.LayoutInflater
import android.widget.Toast
import com.google.firebase.database.*
import com.mrgames13.jimdo.feinstaubapp.R
import com.mrgames13.jimdo.feinstaubapp.model.Sensor
import com.mrgames13.jimdo.feinstaubapp.tool.StorageUtils
import com.mrgames13.jimdo.feinstaubapp.ui.activity.MainActivity
import java.util.*

class WebRealtimeSyncService : Service() {
    internal lateinit var favourites: ArrayList<Sensor>
    internal lateinit var ownSensors: ArrayList<Sensor>

    // Utils packages
    private lateinit var su: StorageUtils
    private var timestamp = 0L

    override fun onBind(intent: Intent): IBinder? = null

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val syncKey = intent.getStringExtra("sync_key")

        // Initialize own instance
        own_instance = this

        // Initialize StorageUtils
        su = StorageUtils(applicationContext)

        // Initialize Firebase
        val db = FirebaseDatabase.getInstance()
        ref = db.getReference("sync/" + syncKey!!)

        refresh(applicationContext)

        return START_NOT_STICKY
    }

    fun refresh(context: Context) {
        timestamp = System.currentTimeMillis()

        // Get favourites and own sensors from local db
        favourites = su.allFavourites
        ownSensors = su.allOwnSensors

        // Assemble data
        val data = HashMap<String, Any>()
        var objectId = 0
        favourites.forEach {
            val sensorMap = HashMap<String, Any>()
            sensorMap["name"] = it.name
            sensorMap["chip_id"] = it.chipID
            sensorMap["color"] = String.format("#%06X", 0xFFFFFF and it.color)
            sensorMap["fav"] = true
            data[objectId.toString()] = sensorMap
            objectId++
        }
        ownSensors.forEach {
            val sensorMap = HashMap<String, Any>()
            sensorMap["name"] = it.name
            sensorMap["chip_id"] = it.chipID
            sensorMap["color"] = String.format("#%06X", 0xFFFFFF and it.color)
            sensorMap["fav"] = false
            data[objectId.toString()] = sensorMap
            objectId++
        }

        // Build connection
        val connection = HashMap<String, Any>()
        connection["time"] = timestamp
        connection["device"] = "app"
        connection["data"] = data
        ref.setValue(connection)

        // Set DateChangeListener
        ref.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snap: DataSnapshot) {
                if (snap.exists()) {
                    if (snap.child("device").value != null && snap.child("device").value != "app") {
                        val newSensors = snap.child("data").value as ArrayList<*>?
                        if (newSensors != null && newSensors.size > 0) {
                            favourites = su.allFavourites
                            ownSensors = su.allOwnSensors
                            favourites.forEach { su.removeFavourite(it.chipID, true) }
                            ownSensors.forEach { su.removeOwnSensor(it.chipID, true) }
                            for (i in newSensors.indices) {
                                val sensor = newSensors[i] as Map<*, *>
                                // Extract data
                                val chipId = sensor["chip_id"].toString()
                                val name = sensor["name"].toString()
                                val favored = sensor["fav"]!!.toString().toBoolean()
                                val color = sensor["color"].toString()
                                if (!su.isFavouriteExisting(chipId) && !su.isSensorExisting(chipId)) {
                                    if (favored) {
                                        su.addFavourite(Sensor(chipId, name, Color.parseColor(color)), true)
                                    } else {
                                        su.addOwnSensor(Sensor(chipId, name, Color.parseColor(color)), offline = true, requestFromRealtimeSyncService = true)
                                    }
                                }
                            }
                            MainActivity.own_instance?.pagerAdapter?.refreshFavourites()
                            MainActivity.own_instance?.pagerAdapter?.refreshMySensors()
                        }
                    }
                } else {
                    ref.removeEventListener(this)
                    // Show toast
                    val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
                    val t = Toast(context)
                    t.setGravity(Gravity.CENTER, 0, 0)
                    t.duration = Toast.LENGTH_LONG
                    t.view = LayoutInflater.from(context).inflate(R.layout.sync_ended, null, false)
                    t.show()
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {
                Toast.makeText(context, "Sync failed", Toast.LENGTH_SHORT).show()
            }
        })
    }

    fun stop() {
        ref.removeValue()
        stopSelf()
    }

    companion object {
        // Variables as objects
        private lateinit var ref: DatabaseReference

        // Variables
        var own_instance: WebRealtimeSyncService? = null
    }
}
